### ISSUE TITLE
preload of already cached resource?

### ISSUE THREAD

**Issue by getify** [w3c/preload#10]
If a resource is already cached, and you try to preload, is the cache checked first? If so, and found, is the "load" event fired synchronously or asynchronously?

**Comment by igrigorik** [comment]
Yes, preload follows the same fetch model as any other request: routed to SW (if one is registered), checked against the HTTP cache, and then routed externally if its a cache miss. 

I'm not sure what "asynchronously" means in this case? Can you elaborate?

**Comment by getify** [comment]
``` js
var i = 1;

var s = document.createElement("link");
s.href = "/path/to/script.js";
s.rel = "preload";
s.policy = "inert";
s.onload = function() { console.log(i); };
document.head.appendChild(s);
i++;
```

Is there ever a chance that `i` printed is `0`, or always `1`?

I ask because:
1. This has been a point of contention in previous preload proposals over the years, as well as other similar loading features, like the script element itself. We need to be explicit about whether the events are sync or async.
2. If it's async, great. Easy. But there's actually a case for the utility of sync, which is not paying the perf penalty for an unnecessary turn of the event loop. IOW, if the resource is already cached, it's already preloaded too, and therefore there's no need to waste the cycle preloading.
   
   If I could do `s.onload = function() { s.loaded = true; }; if (s.loaded) { /* skip */ }` that might be more performant.
   
   Alternatively, if the element itself could have a property like `loaded` that's proactively set (that is, at setting of `href`, the cache hit happens, and `loaded` is set to `true` if present, that would avoid the whole event thing entirely.

**Comment by igrigorik** [comment]
@getify this seems like an orthogonal discussion that's more relevant to internal plumbing of your JS engine then preload. That said... 

The fetch would be initiated when element is added to the document (see https://github.com/w3c/resource-hints/issues/25). You can't assume that cache checks (or SW-related work) is instant, so in practice I guess you'd have to assume that it's "async".

**Comment by getify** [comment]
Assuming it's "async" and knowing for sure it's async are, unfortunately, historically, different things. Think `document.write(..)` with onload events.

Moreover, if we say they are in fact async, is there anyway for me to avoid the cycle of preload in the case where an item is already fresh in cache? That's the much more important point of this thread's line of questioning.

**Comment by igrigorik** [comment]
I guess we should say its "async" then.

> is there anyway for me to avoid the cycle of preload in the case where an item is already fresh in cache?

This assumes that checking the cache is instant, which it's not. This can easily take tens to hundreds of milliseconds.

**Comment by getify** [comment]
> I guess we should say its "async" then.

Thanks. :)

**Comment by yoavweiss** [comment]
> This assumes that checking the cache is instant, which it's not. This can easily take tens to hundreds of milliseconds.

I believe "in cache" here refers to "in memory cache" and testing if something is there is instant and synchronous.

For in memory cache images, the image loading is synchronous, but onload is not. @zcorpan - do you know if the same is true for other external resources?

> Moreover, if we say they are in fact async, is there anyway for me to avoid the cycle of preload in the case where an item is already fresh in cache? That's the much more important point of this thread's line of questioning.

@getify - so what you want here is an API that answers "Is this in memory cache?" If so, IMO that might be out of scope here. If what you're aiming for is "Is this in the HTTP cache?", even more so :)

**Comment by getify** [comment]
@yoavweiss-

I was definitely talking more about "is it in memory" than "is it on disk".

I was just wondering if there could be some way to create a `link rel=preload` element and have a property immediately (synchronously) updated on it if it's "cached" (in memory) already, so you can skip waiting for the "preload" event tick to proceed. That kind of feature is of course optional to the overall design, but it would make it _much_ more compelling to me.

I do not have the desire for a separate API for this... only if there was some way to expose the information on a preload element.

> IMO that might be out of scope here.

I hope not. I hope this is something that streamlines in with the existing design and makes it more useful. It's not a broadly general separate feature or anything like that.

---

**Side Note:** This sort of magical property on the `preload` element could be a clever path for a [feature detect](https://github.com/w3c/preload/issues/7). For example, it could be explicitly set to `true` or `false` (as opposed to just be defined or not) only if preloading was supported. We actually have a subtle trick feature-detect like that for the `async` property on script elements, so it's not at all unprecedented.

---

To the point of my OP, the "load" event is _always_ going to be fired, even if you make a `preload` element for a resource that's already "cached" (in memory). Right?

So if that's always going to happen, being able to know that ahead of time and skip that pretense would be nice.

>  images...the same is true for other external resources?

For my purposes, I think it would be sufficient if at a minimum, scripts and css could be synchronously "cache" checked to avoid preloading. These resources in particular are the main use-case for preloading from my PoV, as they're the ones which affect the page once "executed". Being able to "check" them in advance without those side effects opens up lots of interesting new loading strategies against various performance concerns.

If any other resources _can_ be checked in the same way, that'd be nice of course. But it'd be icing. And if any of those others have some special reason why that can't be done, that's really not a big deal. You'd just pay that slight penalty for those less-commonly-preloaded resource types by waiting an event tick for the event to let you know you can proceed.

**Comment by toddreifsteck** [comment]
@travisleithead pinged me on feature detection of preload last night and his thoughts were very similar to what @getify is suggesting. I'll let Travis reply with his thinking here or on feature detect thread. Same general solution may solve both issues.

**Comment by igrigorik** [comment]
> I was definitely talking more about "is it in memory" than "is it on disk".
> I was just wondering if there could be some way to create a link rel=preload element and have a property immediately (synchronously) updated on it if it's "cached" (in memory) already, so you can skip waiting for the "preload" event tick to proceed. That kind of feature is of course optional to the overall design, but it would make it much more compelling to me.

I agree with @yoavweiss, this is out of scope for preload. The distinction for "in memory" vs. "on disk" vs "somewhere else" is an implementation detail that's UA and architecture specific. Also, we're talking about caches, so "was in memory" is not a signal that it will be in memory.. and then we have ServiceWorker-enabled cases, which we can't say anything meaningful about until/unless we dispatch the request and let it apply its own logic (local copy, synthesize response, remote fetch, etc). 

tl;dr: there is no synchronous signal for "already available" (see https://github.com/w3c/preload/issues/10#issuecomment-78165754).

Closing this; let's continue the feature detect discussion in #7.

**Comment by zcorpan** [comment]
@yoavweiss I think in general we try to make such things async. img is sync because compat requires it (the image cache is special). But e.g. video and fonts are async.
